packages/loopring_v3/contracts/impl/libtransactions/WithdrawTransaction.sol

```diff
@@ -0,0 +1,208 @@
+        "Withdrawal(address owner,uint24 accountID,uint32 nonce,uint16 tokenID,uint256 amount,uint16 feeTokenID,uint256 fee,address to,bytes32 dataHash,uint24 minGas)"
+    );
+
+    struct Withdrawal
+    {
+        uint    withdrawalType;
+        address owner;
+        uint24  accountID;
+        uint32  nonce;
+        uint16  tokenID;
+        uint    amount;
+        uint16  feeTokenID;
+        uint    fee;
+        address to;
+        bytes32 dataHash;
+        uint24  minGas;
+    }
+
+    // Auxiliary data for each withdrawal
+    struct WithdrawalAuxiliaryData
+    {
+        uint gasLimit;
+        bytes signature;
+        bytes auxiliaryData;
+    }
+
+    event ForcedWithdrawalConsumed(
+        uint24  indexed accountID,
+        uint16          tokenID,
+        uint            amount
+    );
+
+    function process(
+        ExchangeData.State storage S,
+        ExchangeData.BlockContext memory ctx,
+        bytes memory data,
+        bytes memory auxiliaryData
+        )
+        internal
+        returns (uint feeETH)
+    {
+        Withdrawal memory withdrawal = readWithdrawal(data);
+
+        WithdrawalAuxiliaryData memory auxData = abi.decode(auxiliaryData, (WithdrawalAuxiliaryData));
+
+        if (withdrawal.withdrawalType == 0) {
+            // Signature checked offchain, nothing to do
+        } else if (withdrawal.withdrawalType == 1) {
+            // Check appproval onchain
+
+            // Calculate the tx hash
+            bytes32 txHash = EIP712.hashPacked(
+                ctx.DOMAIN_SEPARATOR,
+                keccak256(
+                    abi.encode(
+                        WITHDRAWAL_TYPEHASH,
+                        withdrawal.owner,
+                        withdrawal.accountID,
+                        withdrawal.nonce,
+                        withdrawal.tokenID,
+                        withdrawal.amount,
+                        withdrawal.feeTokenID,
+                        withdrawal.fee,
+                        withdrawal.to,
+                        withdrawal.dataHash,
+                        withdrawal.minGas
+                    )
+                )
+            );
+
+            // Verify the signature if one is provided, otherwise fall back to an approved tx
+            if (auxData.signature.length > 0) {
+                require(txHash.verifySignature(withdrawal.owner, auxData.signature), "INVALID_SIGNATURE");
+            } else {
+                require(S.approvedTx[withdrawal.owner][txHash], "TX_NOT_APPROVED");
+                S.approvedTx[withdrawal.owner][txHash] = false;
+            }
+        } else if (withdrawal.withdrawalType == 2 || withdrawal.withdrawalType == 3) {
+            // Forced withdrawals cannot make use of certain features because the
+            // necessary data is not authorized by the account owner.
+            require(withdrawal.owner == withdrawal.to, "INVALID_WITHDRAWAL_ADDRESS");
+            require(withdrawal.fee == 0, "FEE_NOT_ZERO");
```

Comment:
For a valid forced withdrawal, will the operator be able to treat it as an "invalid" one by purposely using an bad 'owner' and 'to' in the data? If so, all forced withdrawals are not forced at all.

Reply:
The operator has not control over `owner` because it's the value stored for the account in the Merkle tree.

Reply:
@yueawang @hosschao we need to pay attention to the corresponding logic in the circuit.!!!